### ISSUE TITLE
Feature request: allow setting syntax in .stylelintrc

### ISSUE THREAD

**Issue by chrisvfritz** [stylelint/stylelint#2973]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Feature request: new option.

> Which version of stylelint are you using?

`8.2.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Via editor extensions such as [vscode-stylelint](https://github.com/shinnn/vscode-stylelint), which uses the Node API under the hood.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes.

> What did you expect to happen?

I would like a `syntax` option in `.stylelintrc`, which would be respected by the CLI, PostCSS plugin, Node API, etc. For example:

``` js
"syntax": "scss"
```

would be equivalent to:

```
--syntax scss
```

on the CLI.

> What actually happened (e.g. what warnings or errors you are getting)?

When using processors to parse alternative syntaxes nested inside other files (e.g. SCSS in a `.vue` file), the correct syntax can't be detected from the file extension. The result is an automatic fallback to CSS, which leads to syntax errors.

You can get around this when you have direct access to the Stylelint runner (e.g. `--syntax scss` on the CLI), but when using a 3rd-party tool that only reads the config, there's no way to pass it the correct syntax.

---

I'm happy to submit a PR, but wanted to check in here first to make sure it'd be welcome.

**Comment by jeddy3** [comment]
@chrisvfritz Thanks for the request and for using the template.

I just took a trip down memory lane trying to work out why we didn't just stick the syntax choice in the configuration object in the first place. The `--syntax` option was added two years and two days ago in https://github.com/stylelint/stylelint/pull/419, and I don't think the `<style>` use case was on our radar at the time.

This feature did get surfaced in https://github.com/stylelint/stylelint/issues/1749, but I think the discussion about moving the syntax choice to the configuration object got sidelined by the discussion on how stylelint should support different syntaxes.

The `--custom-syntax` option was then added in https://github.com/stylelint/stylelint/pull/2017. This option was added to allow users to use a syntax not bundled with stylelint (and to provide a distinction between "default" syntaxes and custom syntaxes).

Long story short, I can't think of any reason why we can't accommodate this request now.(@stylelint/core Can anyone else?) We'll need to decide what happens when the syntax is set both within the configuration object and via an option. Would the `--syntax` option taking precedence over `syntax` in the configuration object be the most intuitive approach?

We'll probably want to add `customSyntax` to the configuration object too (and this would take precedence over `syntax` if they are both set), but I think we can cross that bridge if that specific feature is ever requested.

> I would like a syntax option in .stylelintrc, which would be respected by the CLI, PostCSS plugin, Node API, etc.

I believe it can be respected by the stylelint CLI and stylelint Node API, but not by the stylelint PostCSS plugin as the syntax choice [is set](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/css-processors.md#parsing-non-standard-syntax) via PostCSS's `syntax` option.

> I'm happy to submit a PR, but wanted to check in here first to make sure it'd be welcome.

That's very kind, thanks. Let's wait to see if anyone else remembers anything pertinent about the `--syntax` option, and if nothing comes up then a PR would be very welcomed.

**Comment by chrisvfritz** [comment]
Thanks for the detailed response!

> We'll need to decide what happens when the syntax is set both within the configuration object and via an option. Would the --syntax option taking precedence over syntax in the configuration object be the most intuitive approach?

As a user, that's what I'd expect.

> Let's wait to see if anyone else remembers anything pertinent about the --syntax option, and if nothing comes up then a PR would be very welcomed.

Sounds great. 🙂  Just let me know when I should get started.

**Comment by gucong3000** [comment]
@chrisvfritz 
Here is a PR for support `.vue` file. #2975

**Comment by chrisvfritz** [comment]
@gucong3000 Excellent! 😄  I currently do have it working in `.vue` files using:

``` js
processors: [
  [
    '@mapbox/stylelint-processor-arbitrary-tags',
    {
      // Prevents non-Vue files from being parsed as HTML
      fileFilterRegex: [/\.vue$/]
    }
  ]
],
extends: [
  'stylelint-config-standard',
  'stylelint-config-css-modules'
],
rules: [
  // Prevents empty .vue files from breaking
  'no-empty-source': null
]
```

Does that PR mean I would no longer need to use a processor?

**Comment by gucong3000** [comment]
> Does that PR mean I would no longer need to use a processor?
Yes.

**Comment by chrisvfritz** [comment]
Wow. That sounds like it would solve my use case _and more_ then. 😍

**Comment by gucong3000** [comment]
@chrisvfritz 
If your using stylelint for vue, you can take a look for this PR: https://github.com/vuejs-templates/webpack/pull/842

**Comment by chrisvfritz** [comment]
Done. 🙂

**Comment by jeddy3** [comment]
Closing as the specific use case was addressed in https://github.com/stylelint/stylelint/pull/2975
